### PR TITLE
Development

### PR DESCRIPTION
**Pull request by madcido**

### CODE REVIEW

File: db/schema.rb [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+  create_table "events", force: :cascade do |t|
+    t.string "name"
+    t.text "description"
+    t.date "date"
+    t.datetime "created_at", null: false
+    t.datetime "updated_at", null: false
+    t.integer "creator_id"
+    t.boolean "private"
+    t.index ["creator_id"], name: "index_events_on_creator_id"
```
Comment (rvvergara) [inline db/schema.rb:32]: We need to add foreign_key constraint on `creator_id`

File: db/schema.rb [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+  create_table "invitations", force: :cascade do |t|
+    t.integer "event_id"
+    t.integer "invited_id"
+    t.integer "invitor_id"
+    t.datetime "created_at", null: false
+    t.datetime "updated_at", null: false
+    t.index ["event_id"], name: "index_invitations_on_event_id"
+    t.index ["invited_id"], name: "index_invitations_on_invited_id"
+    t.index ["invitor_id"], name: "index_invitations_on_invitor_id"
```
Comment (rvvergara) [inline db/schema.rb:43]: We need foreign_key constraints for:
`invitor_id`
`invited_id`
`event_id`

File: db/schema.rb [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+# database schema. If you need to create the application database on another
+# system, you should be using db:schema:load, not running all the migrations
+# from scratch. The latter is a flawed and unsustainable approach (the more migrations
+# you'll amass, the slower it'll run and the greater likelihood for issues).
+#
+# It's strongly recommended that you check this file into your version control system.
+
+ActiveRecord::Schema.define(version: 2019_05_21_004956) do
+
+  create_table "attendances", force: :cascade do |t|
```
Comment (rvvergara) [inline db/schema.rb:15]: We need foreign_key constraints for:
`user_id`
`event_id`

File: app/models/event.rb [hunk 4]
```diff
@@ -0,0 +1,11 @@
+class Event < ApplicationRecord
```
Comment (rvvergara) [inline app/models/event.rb:1]: Let us put validations at least on the presence of event name and date. Right now I can create an empty event.

File: app/models/user.rb [hunk 5]
```diff
@@ -0,0 +1,22 @@
+class User < ApplicationRecord
+    has_many :events, foreign_key: "creator_id"
+
+    has_many :attendances
+    has_many :attended_events, through: :attendances, source: "event"
+
+    has_many :created_invites, class_name: "Invitation", foreign_key: "invitor_id"
```
Comment (rvvergara) [inline app/models/user.rb:7]: To have a more meaningful association, I suggest we do away with this particular one altogether. I see that you are using this in your invitations_controller. The better way to create an invitation is anchor it on the `invitation` itself and not on the user. See my comment on the `invitations_controller.rb` file

File: app/controllers/invitations_controller.rb [hunk 6]
```diff
@@ -0,0 +1,20 @@
+class InvitationsController < ApplicationController
+        
+    def create
+        @invitation = current_user.created_invites.build(invitation_params)
```
Comment (rvvergara) [inline app/controllers/invitations_controller.rb]: We want the creation of the invitation to anchor more on the invitation itself than the `current_user`. In this case we can simply do:
`@invitation = invitation.create(invitation_params)`

Your `invitation_params` should then include `current_user` 's id. Such that the full parameters you will get from your form are:
`event_id`
`current_user.id`
`invited_id`

File: app/views/events/index.html.erb [hunk 7]
```diff
@@ -0,0 +1,26 @@
+<div class="centered">
+    <h1>Upcoming Events</h1>
+    <% @upcoming_events.each do |event| %>
```
Comment (rvvergara) [inline app/views/events/index.html.erb]: We are using `each` heavily. There is obvious repetition in `@upcoming_events` and `@past_events`

We should have an `_event.rb` partial that we will render instead.

File: app/views/events/index.html.erb [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+            <p><%= event.description %></p>
+            <p>Created by: <%= link_to event.creator.name, event.creator %></p>
+        </div>
+        <hr>
+    <% end %>
+</div>
+<br>
+<div class="centered">
+    <h1>Past Events</h1>
+    <% @past_events.each do |event| %>
```
Comment (rvvergara) [inline app/views/events/index.html.erb]: Let's use an `_event.rb` partial

File: app/views/events/show.html.erb [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+    <p>Description: <%= @event.description %></p>
+    <hr>
+    <div class="row space-between">
+        <p>Date: <%= @event.date %></p>
+        <% if @event.private? %>
+            <p>Type: Private</p>
+        <% else %>
+            <p>Type: Open</p>
+        <% end %>
+        <p>Creator:  <%= link_to @event.creator.name, @event.creator %> </p>
```
Comment (rvvergara) [inline app/views/events/show.html.erb]: To make this `@event.creator.name` call simpler we could have created a `@creator` instance variable in the controller (@creator = @event.creator).

File: app/views/events/show.html.erb [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+    <div class="row space-around">
+        <div class="att-inv">
+            <% if !@event.private? || @invitation %>
+                <p class="center">Take part?</p>
+                <% if @event.attendees.include?(current_user) %>
+                    <%= form_for @attendance, method: :delete, url: attendance_path(@event.attendances.find_by(user_id: current_user.id)) do |f| %>
+                        <%= f.submit "Leave Attendance" %> 
+                    <% end %>
+                <% else %>
+                    <%= form_for @attendance do |f| %>
```
Comment (rvvergara) [inline app/views/events/show.html.erb]: Why didn't we just use a partial for the form?

File: app/views/events/show.html.erb [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+                <% @event.attendees.each do |attendee| %>
+                    <li> <%= link_to attendee.name, user_path(attendee.id) %> </li>
+                <% end %>
+            </ul>
+        </div>
+        <% if @event.private? %>
+            <div class="att-inv">
+                <% if @invitation %>
+                    <p class="center">Invite people!</p>
+                    <%= form_for @new_invitations do |f| %>
```
Comment (rvvergara) [inline app/views/events/show.html.erb]: Let's use a partial for this form

File: app/views/events/show.html.erb [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+    <% if @invitation && @invitation.invitor != current_user %>
+        <p class="center">You have been invited to this event by <%= @invitation.invitor.name %> </p>
+    <% end %>
+
+    <div class="row space-around">
+        <div class="att-inv">
+            <% if !@event.private? || @invitation %>
+                <p class="center">Take part?</p>
+                <% if @event.attendees.include?(current_user) %>
+                    <%= form_for @attendance, method: :delete, url: attendance_path(@event.attendances.find_by(user_id: current_user.id)) do |f| %>
```
Comment (rvvergara) [inline app/views/events/show.html.erb]: Why didnt we use a form partial?

**Review by rvvergara — CHANGES_REQUESTED** [review]
Hi @madcido and @danmontielh,

Dayyym, nice UI guys! So freakin awesome! I especially like the sliding 'New Event' form. Just one thing about the UI, if we can display a more readable date format. So instead of `2019-05-25` let's display `May 25, 2019`.

In the user display pages (by the way since you have a route for user show page you should put a link to all users) what do you mean by 'Past Events'? Are these past events the user has attended or created? It isn't clear from the page what this means.

What is the mechanism to invite another user to an event? All I could see on a user's page is 'Invitations not accepted'. 

Another thing is it would be great if users are not able to click `Attend this Event` for past events.

Please see further my comments on some of the models and controllers

**Comment by madcido** [comment]
@rvvergara 
Thanks for the feedback. About the invitations mechanism, it works as follows:
- users can attend open events as desired
- users can only attend private events if invited
- only users already invited to an event can invite other users to this particular event
- when a user is invited to an event, this event will be shown in the "invitations not accepted" list and this user is free to attend or not to this event (as if it were open)
- if the user decides to attend the invited event, the event will disappear from the "invitations not accepted" list

You can see it working if you create two different users and a private event. We also thought about creating a notification system to let users know they just got invited to something (besides the "invitations not accepted" list), but... maybe in the future.

I just have two questions for you: 
1 - Besides readability, any particular reason to use partials for every form? (since they are not going to be re-used in any other template, I just didn't think partials were necessary)
2 - Any technical reason to anchor invitations on themselves?

File: db/schema.rb [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+    t.index ["invitor_id"], name: "index_invitations_on_invitor_id"
+  end
+
+  create_table "users", force: :cascade do |t|
+    t.string "name"
+    t.datetime "created_at", null: false
+    t.datetime "updated_at", null: false
+  end
+
+end
```
Comment (rvvergara) [inline db/schema.rb:52]: It appears indeed there's a bug when we are using `sqlite3` in Rails 5. So I can safely assume foreign_keys were definitely added (albeit silently by sqlite3).

For succeeding projects then I highly recommend we use postgresql instead.

File: app/models/event.rb [hunk 14]
```diff
@@ -0,0 +1,13 @@
...
+    
+    has_many :attendances
+    has_many :attendees, through: :attendances, source: "user"
+    
+    has_many :invitations
+    has_many :inviteds, through: :invitations
+
+    default_scope -> { order(date: :desc) }
+
+    validates :name, :date, presence: true
```
Comment (rvvergara) [inline app/models/event.rb:10]: Awesome!

File: app/models/event.rb [hunk 15]
```diff
@@ -0,0 +1,13 @@
+class Event < ApplicationRecord
+    belongs_to :creator, class_name: "User"
```
Comment (rvvergara) [inline app/models/event.rb:2]: When the event is created, it's good that the event creator is automatically an attendee. However, I notice that the event creator also automatically has an 'invite' record. How about removing an outstanding invite once the user is already an attendee?

File: app/views/events/index.html.erb [hunk 16]
```diff
@@ -0,0 +1,2 @@
+<%= render "event", collection: @upcoming_events, heading: "Upcoming Events" %>
```
Comment (rvvergara) [inline app/views/events/index.html.erb]: This is better. But I wanna suggest this:

```
<div class="centered">
  <h1>Upcoming</h1>
  <%= render "event", collection: @upcoming_events, heading: "Upcoming Events" %>
</div>

<div class="centered">
  <h2>Past Events</h2>
  <%= render "event", collection: @past_events, heading: "Past Events" %>
</div>

```
We still want the headings in the index page to make this page more understandable.

**Review by rvvergara — CHANGES_REQUESTED** [review]
Hi @madcido and @danmontielh,

Great questions! 

Ok here are my thoughts:

1. Regarding using partials even if you're not using `<form>` code in other files:
 - Code can be quite hard to debug especially if we have too much stuff going on in a single file. There are 2 types of people we are writing our code for - a) For our app's actual users and b) for fellow developers who may one day either collaborate with us or take over in extending our apps. For the former, we want them to have the best user experience possible. In this I believe you are doing an excellent job (except for the clarifications in the UI I previously requested). For the latter group, we want our code to be very readable and easily understandable for them (and to our future selves as well, should we revisit this project and wanna extend it). Rails is a convention-centric framework in that it has as a core principle easy collaboration. That's my long way of saying putting the form in its own partial will make it easier later on for somebody to reason about with your project (including your future selves). 

2. Regarding anchoring invitation objects onto the `invitation` model -> This is more of a design choice, like the form partial above. Technically speaking, it's pretty obvious any way works. We do need to think about 'single responsibility'. The user model isn't responsible for creating invitations, it is the invitation that is responsible for that. We wanna be explicit about it so we should leave all the creation of invites to the invitation model itself. Again keep in mind collaborators and your future self. If another person looks at your code it's easier to understand if he/she sees the creation of invitations expressed in terms of the invitation and not the user. 

But yes we can agree on either way of writing it. After all, for a blog project, for instance, we write `user.posts.build(post_params)` to create a post. So pls take what I previously said as a suggestion. Though I know it would make your code much better.

## User Interface 

Ok in relation to inviting a user. I will relent on this. Though I would recommend having a button in the event page that allows the event creator to invite other users (possibly a list of other users would come in handy in the event page, such that the creator can click on an "Invite" button beside each user?).

Another thing, let's make the sooner happening upcoming events listed ahead of the ones happening after it. Right now the later ones are listed first. We want users to first see the ones happening sooner.

See my further reviews in some of the other files

**Comment by madcido** [comment]
> Though I would recommend having a button in the event page that allows the event creator to invite other users (possibly a list of other users would come in handy in the event page, such that the creator can click on an "Invite" button beside each user?).

@rvvergara 
I'm not sure if I understand you right, since there's already a button and a list in the event page to invite other users (only for private events, but it is there). Anyway, thanks for the other ideas, they were indeed great improvement.

**Review by rvvergara — APPROVED** [review]
Hi @madcido and @danmontielh,

I think everything's good now! yeah well the last comment I made regarding the 'invite' button in the event show page because i don't see that button:

image

Like in this page, there's another user `Mike` that i created but i don't see his name anywhere nor a button that lets me invite anybody else.

But anyway it's a UI thing. Though it's great to refine this, I guess you've already satisfied the main requirements of the project. So you may already merge this into `master`

# :+1:

**Pull request merged**